Heavy ion Vs algorithm same event subtraction

yslai opened the pull request:
A change to implement two features:

(a) A "fall-back" mode, when no calibration information is available, that estimates the UE from the same event using a close approximation of the ATLAS algorithm [Physics Letters B 719, 220–241 (2013)]. This mode is not turned on by default.

(b) Flags to expose the Fourier orders that are going to be subtracted. This was inadvertently hard-coded before.

The changes are merged due to otherwise conflicting diffs.

Commit:
HF/Voronoi self subtraction plus vn controlling flags

Commit:
HF/Voronoi self subtraction plus vn controlling flags

Commit:
HF/Voronoi self subtraction plus vn controlling flags

Commit:
HF/Voronoi self subtraction plus vn controlling flags

cmsbuild commented:
A new Pull Request was created by @yslai  for CMSSW_7_6_X.

Vs same event vn flags 3

It involves the following packages:

RecoHI/HiJetAlgos


@cmsbuild, @cvuosalo, @slava77 can you please review it and eventually sign? Thanks.
@jazzitup, @dgulhan, @yenjie, @kurtejung, @istaslis, @mandrenguyen, @richard-cms, @yetkinyilmaz this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
The jenkins tests job failed, please try again.
https://cmssdt.cern.ch/jenkins/job/ib-any-integration/8342/console

cmsbuild commented:
The jenkins tests job failed, please try again.
https://cmssdt.cern.ch/jenkins/job/ib-any-integration/8343/console

cmsbuild commented:
The jenkins tests job failed, please try again.
https://cmssdt.cern.ch/jenkins/job/ib-any-integration/8344/console

cmsbuild commented:
The jenkins tests job failed, please try again.
https://cmssdt.cern.ch/jenkins/job/ib-any-integration/8345/console

yslai commented:
From the log:

OSError: [Errno 28] No space left on device: '1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15'

Is this really a problem from my side?

cmsbuild commented:
+1
Tested at: 5c427da5340574133c85b7a46ef7031af6ec7e03
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11452/8346/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11452/8346/summary.html

cvuosalo commented:
@yslai:
Please change the title of this PR to be a plain English description and include "Heavy Ions" in the title.

Also, fix the line in the description that says "expose the to be subtracted".

cvuosalo commented:
The file BPMPDInterface.h is not a proper header file because it contains complex code. Please remove the implementations of complex methods to a new BPMPDInterface.cc. The implementation of a method should only appear in the header file if it is trivial.

cvuosalo commented on RecoHI/HiJetAlgos/plugins/BPMPDInterface.h:
```diff
@@ -0,0 +1,848 @@
+			}
+			else {
+				fprintf(stderr, "%s:%d: illegal sense (`%s')\n",
+						__FILE__, __LINE__, constraint_sense.c_str());
```
fprintf(stderr...) is not allowed. Please convert to use the standard LogError().

yslai replied:
Fixed with the new commit

cvuosalo commented on RecoHI/HiJetAlgos/plugins/BPMPDInterface.h:
```diff
@@ -0,0 +1,848 @@
+				return 500;
+			}
+			fprintf(stderr, "%s:%d: %d\n", __FILE__, __LINE__,
+					termination_code);
```
Please convert all fprintf to LogError().

yslai replied:
Fixed with the new commit.

cvuosalo commented on RecoHI/HiJetAlgos/plugins/GenericVoronoiAlgorithm.h:
```diff
@@ -0,0 +1,143 @@
+	size_t _ncost;
+	std::vector<size_t> _nblock_subtract;
+	void *_lp_environment;
+	void *_lp_problem;
```
Where is _lp_environment set? And where is _lp_problem used?

yslai replied:
This became a scoped variable and is inadvertently left in the class declaration. Fixed with the new commit.

cvuosalo commented on RecoHI/HiJetAlgos/plugins/SameEventVoronoiAlgorithm.cc:
```diff
@@ -0,0 +1,399 @@
+	{
+		// binary search in the array of towers eta edges
+
+		int size = 42;
```
Use sizeof(etaedge)/sizeof(double) rather than 42 here.

yslai replied:
Fixed with the new commit.

cvuosalo commented on RecoHI/HiJetAlgos/plugins/SameEventVoronoiAlgorithm.cc:
```diff
@@ -0,0 +1,399 @@
+
+			if( curr >= size || curr < 1 )
+				std::cout <<  " ParticleTowerProducer::eta2ieta - wrong current index = "
+						  << curr << " !!!" << std::endl;
```
Please convert all cout to LogError().

yslai replied:
Fixed with the new commit.

cvuosalo commented on RecoHI/HiJetAlgos/plugins/SameEventVoronoiAlgorithm.cc:
```diff
@@ -0,0 +1,399 @@
+	}
+	std::transform(density_0.begin(), density_0.end(), density_2.begin(),
+				   std::bind1st(std::multiplies<double>(),
+								density_2_sum / density_0_sum));
```
Please check density_0_sum > 0 before dividing by it.

yslai replied:
This is now checked with the new commit.

cvuosalo commented on RecoHI/HiJetAlgos/plugins/SameEventVoronoiAlgorithm.cc:
```diff
@@ -0,0 +1,399 @@
+
+			jet_excluded &= (et_max >= _same_event_fake_reject_et_max &&
+							 et_max / et_mean >=
+							 _same_event_fake_reject_et_max_over_mean);
```
More protection against division by 0 needed. Please check pseudotower.size() > 0 and et_mean > 0.

yslai replied:
This is now checked with the new commit.

cvuosalo commented on RecoHI/HiJetAlgos/plugins/VoronoiBackgroundProducer.cc:
```diff
@@ -143,7 +162,12 @@ VoronoiBackgroundProducer::produce(edm::Event& iEvent, const edm::EventSetup& iS
+	}
+	else {
+	voronoi_ = new VoronoiAlgorithm(ue,equalizeR_,excludeV1_,maxVn_,diagonalVn_,std::pair<double, double>(equalizeThreshold0_,equalizeThreshold1_),doEqualize_);
+	}
```
Is voronoi_ ever deleted? It might be a good idea to convert it into a unique_ptr to protect against memory leaks.

yslai replied:
Changed to unique_ptr, as suggested, with the new commit.

cvuosalo commented:
Please fix the problems found in your code by the static analyzer:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11452/8346/llvm-analysis/

cvuosalo commented:
@yslai:
When might this PR be updated?

lgray committed:
port HGCalRecAlgos/Producers in from CMSSW_6_2_0_SLHC26_patch3

lgray committed:
revive testHGCalLocalReco_cfg.py

Commit:
add explicit overflow check for input regions to jet finder

fixes non-overflowed forward jets

Commit:
do not apply background subtraction to overflow regions

Commit:
MHT fixes for 74X

cvuosalo commented:
@yslai: Time is running out for 76X. Prompt action to fix the issues in this PR is needed to improve its chances of getting into 76X.

lgray committed:
use values from the digitizer directly

lihux25 committed:
Fix the duplicated HF channels. Throw a LogError is this is found and an empty HF digi collection is put into event

lgray committed:
address code review

lgray committed:
fix clang error

lgray committed:
kill clang warning

Dr15Jones committed:
Removed unused DetSetLazyVector class

The class DetSetLazyVector is not used anywhere in CMSSW and it is
not const thread-safe. Therefore it has been removed.

Dr15Jones committed:
Removed unused typedef from HLTTrackHaloFilter

The typedef was to a thread unsafe class. The typedef was not being
used so removing it is harmless and it avoids false positives when
searching for thread-safety issues.

cmsbuild committed:
Merge pull request #11720 from fwyzard/retire_cmsCheckMultithreading_76x

cmsCheckMultithreading replaced by edmCheckmultithreading

cmsbuild committed:
Merge pull request #11704 from lihux25/fix_HF_dup_chns

Fix the duplicated HF channels

cmsbuild committed:
Merge pull request #11676 from richard-cms/L1_MHTandForJetFixes_76X

L1 MHT and forward Jet fixes

cmsbuild committed:
Merge pull request #11714 from Dr15Jones/removeUnusedTypedef

Removed unused typedef from HLTTrackHaloFilter

cmsbuild committed:
Merge pull request #11713 from Dr15Jones/removeDetSetLazyVector

Removed unused DetSetLazyVector class

cmsbuild committed:
Merge pull request #11651 from lgray/hgcal_rechit_modules

Forward port HGCalRecAlgos/Producers from CMSSW_6_2_0_SLHC26_patch3 (76X)

Commit:
Merge branch 'vs_same_event_vn_flags_3' of https://github.com/yslai/cmssw into from-CMSSW_7_6_X_2015-10-12-1100

Commit:
PR comments

Commit:
PR comments

cmsbuild commented:
Pull request #11452 was updated. @perrotta, @smuzaffar, @Dr15Jones, @cvuosalo, @fwyzard, @Martin-Grunewald, @cmsbuild, @slava77, @mulhearn can you please check and sign again.

yslai commented:
My apologies with the delay, being distracted with other tasks. I have implemented the fixes from the comment with a new commit.

Martin-Grunewald commented:
-1
please rebase properly without pulling in other PRs.

Commit:
Merge branch 'vs_same_event_vn_flags_3' of https://github.com/yslai/cmssw into vs_same_event_vn_flags_3

cmsbuild commented:
Pull request #11452 was updated. @perrotta, @smuzaffar, @Dr15Jones, @cvuosalo, @fwyzard, @Martin-Grunewald, @cmsbuild, @slava77, @mulhearn can you please check and sign again.

yslai commented:
Abandoned for a new, clean PR
